### ISSUE TITLE
Keep/build old package versions in MELPA Stable

### ISSUE THREAD

**Issue by purcell** [melpa/melpa#4712]
If we start building/keeping old package versions in MELPA Stable, this may drive  upstream (package.el / 3rd-party-extension) support for pinning.

It's not a trivial change, since `:files` specs in recipes describe the upstream repo's _current_ contents, not its shape at some older point in time.

Additionally, this will require more build time and disk space, and therefore cost.

See https://github.com/melpa/melpa/issues/1749#issuecomment-294349264

**Comment by Fuco1** [comment]
What I would do first is just keep the index; you need to keep the dependencies of each package in each version, but this is significantly less than having all the packages built and archived.  That can then be done on-demand, probably.

It would also be amazing if the recipe was stored in the repository itself, something like a Cask file (or package.el file or something like that), and the melpa recipe would be just a pointer to the repo.  This would solve the problem about past version (well, not retroactively, but for all future releases at least).

That seems like how it should've been from the start, but I understand that would have created a "bootstrap" problem where people would just be unwilling to add the manifest to their packages.  Nowadays it seems mode doable, especially with people getting used to all the other package managers that work like that.

**Comment by purcell** [comment]
> That can then be done on-demand, probably.

Not currently: we build in a batch and serve everything statically.

> It would also be amazing if the recipe was stored in the repository itself, something like a Cask file (or package.el file or something like that), and the melpa recipe would be just a pointer to the repo. This would solve the problem about past version (well, not retroactively, but for all future releases at least).

Yep, I'd be open to this: Cask just wraps `package-build` these days anyway. We'd need a way of reporting on broken builds, those, so that when people break their recipes, it's obvious.

> That seems like how it should've been from the start, but I understand that would have created a "bootstrap" problem

Exactly. And I think you'd still find that we'd be lucky to get more than 25% of packages to provide their own recipes upstream, though if they're the most popular 25%, that's pretty good!

**Comment by preetpalS** [comment]
> It would also be amazing if the recipe was stored in the repository itself, something like a Cask file (or package.el file or something like that), and the melpa recipe would be just a pointer to the repo. This would solve the problem about past version (well, not retroactively, but for all future releases at least).

Maybe packages could be allowed to opt-in to storing the recipe within their own repository. Here are three potential ways this could occur:

1. Add another section to the comment metadata to allow for the opt-in to occur.

For example:

```elisp
;; Author: First Last
...
;; MELPA_SELF_HOSTED_RECIPE_OPT_IN: true
License: ...
```

2. Allow for opting-in via placing a specially named file in the repository.
3. Allow for opting-in via the creation of a special branch in the repository (like how GitHub repository specific sites work).

Once the opt-in occurs (once it is detected by the build process, possibly followed by a successful build), maybe the build process can remove the existing recipe and replace it with only a pointer to the repository. Additionally once MELPA has migrated the package metadata (recipe) to only be a pointer to the package's repository, the package authors would be free to remove the then unnecessary opt-in metadata from the their repository.

**Comment by alphapapa** [comment]
Sorry, I'm a bit confused: do you mean to move the non-stable version of a package to the Stable repo when a package in the non-stable is updated?

If so, this seems strange to me, because such a package will not necessarily be "stable," merely old.  In fact, since it's now obsolete, it's likely to be less "stable" than the version in the non-stable repo, because it's likely to have bugs that have been fixed.

I think it would be great to have an "archive" repo, similar to what snapshot.debian.net does for Debian, but it should probably be called that.

Or maybe I'm misunderstanding what you mean...

**Comment by purcell** [comment]
> @purcell Sorry, I'm a bit confused: do you mean to move the non-stable version of a package to the Stable repo when a package in the non-stable is updated?

No. I mean that MELPA Stable currently only contains the latest-numbered version of a package, so when the number is bumped, the older package disappears.
